horqrux/circuit.py

```diff
@@ -1,47 +1,51 @@
 from __future__ import annotations
 
+from collections import Counter
 from dataclasses import dataclass
 from typing import Any, Callable
 from uuid import uuid4
 
+import jax
+import jax.numpy as jnp
 from jax import Array
 from jax.tree_util import register_pytree_node_class
 
+from horqrux.adjoint import ad_expectation, adjoint_expectation
 from horqrux.apply import apply_gate
 from horqrux.parametric import RX, RY, Parametric
 from horqrux.primitive import NOT, Primitive
-from horqrux.utils import zero_state
+from horqrux.utils import DiffMode, zero_state
 
 
 @register_pytree_node_class
 @dataclass
-class Circuit:
+class QuantumCircuit:
     """A minimalistic circuit class to store a sequence of gates."""
 
     n_qubits: int
-    feature_map: list[Primitive]
-    ansatz: list[Primitive]
+    operations: list[Primitive]
 
-    def __post_init__(self) -> None:
-        self.state = zero_state(self.n_qubits)
-
-    def __call__(self, param_values: Array) -> Array:
-        return apply_gate(
-            self.state,
-            self.feature_map + self.ansatz,
-            {name: val for name, val in zip(self.param_names, param_values)},
-        )
+    def __call__(self, state: Array, values: dict[str, Array]) -> Array:
+        if state is None:
+            state = zero_state(self.n_qubits)
+        return apply_gate(state, self.operations, values)
 
     @property
     def param_names(self) -> list[str]:
-        return [str(op.param) for op in self.ansatz if isinstance(op, Parametric)]
+        return [
+            str(op.param)
+            for op in self.operations
+            if isinstance(op, Parametric)
+            and isinstance(op.param, str)
+            and op.param.startswith("v_")
```

Comment:
Are these two checks necessary ?

Reply:
Hm I guess what was left back then was a way to differentiate between variational parameters and non variational parameters. Let me examine this again later.

Reply:
@RolandMacDoland decided to add a field to separate feature map parameters from other variational parameters.